ipykernel/kernelbase.py

```diff
@@ -810,16 +790,14 @@ def _topic(self, topic):
 
     _aborting = Bool(False)
 
-    @gen.coroutine
-    def _abort_queues(self):
+    async def _abort_queues(self):
         self.shell_stream.flush()
         self._aborting = True
-
-        def stop_aborting(f):
+        def stop_aborting():
             self.log.info("Finishing abort")
             self._aborting = False
-
-        self.io_loop.add_future(gen.sleep(self.stop_on_error_timeout), stop_aborting)
+        await asyncio.sleep(self.stop_on_error_timeout)
```

Comment:
This changes behavior substantially - the original behavior was to stop aborting after a short delay, to flush the shell queue, essentially.

This change means that whoever calls `_abort_queues` blocks until this is done. If that is in the same thread as the shell channel message handling, then nothing really happens to the queue, as messages waiting in the queue don't get cancelled. (This affects the stop_on_error behavior in cases like restart & run all)

To preserve the same behavior, use `call_later(self.stop_on_error_timeout, stop_aborting)`

Reply:
> `call_later(self.stop_on_error_timeout, stop_aborting)`

That's what I originally did, however this was breaking the tests.

Reply:
> This change means that whoever calls _abort_queues blocks until this is done.

Actually I don't think that this is true. It will only block for the amount of time specified in `stop_on_error_timeout`.

Reply:
> It will only block for the amount of time specified in stop_on_error_timeout.

That's true, but it prevents the flushing that should be occurring while we wait for this timeout. The point of this timeout is to abort any requests we receive during this window. What does the failure look like when the original behavior is preserved?

I suspect this might be working in tests because the control thread doesn't block the main thread by default.

Aside: What would be even better would be to check the request date header and compare with clock time, but that's vulnerable to hard-to-diagnose clock drift, so probably unreliable.

Reply:
Fixed. I don't know what was wrong earlier with that.